pkg/snapshot/restorer/restorer.go

```diff
@@ -317,24 +326,260 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
 	return e, nil
 }
 
-// applyDeltaSnapshot applies the events from time sorted list of delta snapshot to etcd sequentially
-func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, snapList snapstore.SnapList) error {
+// applyDeltaSnapshots fetches the events from delta snapshots in parallel and applies them to the embedded etcd sequentially
+func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, ro RestoreOptions) error {
+	snapList := ro.DeltaSnapList
+	numMaxFetchers := ro.MaxFetchers
+
+	// no delta snapshots available
+	if len(snapList) == 0 {
+		return nil
+	}
+
 	firstDeltaSnap := snapList[0]
+
+	r.logger.Infof("Applying delta snapshot %s", path.Join(firstDeltaSnap.SnapDir, firstDeltaSnap.SnapName))
 	if err := r.applyFirstDeltaSnapshot(client, *firstDeltaSnap); err != nil {
 		return err
 	}
 	if err := verifySnapshotRevision(client, snapList[0]); err != nil {
 		return err
 	}
-	for _, snap := range snapList[1:] {
-		if err := r.applyDeltaSnapshot(client, *snap); err != nil {
-			return err
+
+	// no more delta snapshots available
+	if len(snapList) == 1 {
+		return nil
+	}
+
+	var (
+		remainingSnaps         = snapList[1:]
+		numSnaps               = len(remainingSnaps)
+		snapLocations          = make([]string, numSnaps)
+		errCh                  = make(chan error, 1)
+		snapInfoCh             = make(chan snapInfo, numSnaps)
+		fetchedEventsInfoCh    = make(chan eventsInfo, numSnaps)
+		applyInfoCh            = make(chan applyInfo, numSnaps)
+		numFetchers            = int(math.Min(float64(numMaxFetchers), float64(numSnaps)))
+		stopCh                 = make(chan bool, numFetchers+2) // used for stopping all routines (fetchers, persister and applier)
+		fetchersStoppedCh      = make(chan bool, numFetchers)   // used to indicate that fetchers have stopped
```

Comment:
It could be simplified with use of [`WaitGroup`](https://golang.org/pkg/sync/#WaitGroup). What do you think?

Reply:
Yes. I've implemented this now, and eliminated `fetchersStoppedCh`, `persistDeltasStoppedCh` and `applyStoppedCh`. Thanks.